gematria/datasets/python/exegesis_benchmark.cc

```diff
@@ -0,0 +1,70 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string_view>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "gematria/datasets/exegesis_benchmark_lib.h"
+#include "gematria/datasets/find_accessed_addrs.h"
+#include "gematria/llvm/llvm_to_absl.h"
+#include "llvm-c/Target.h"
+#include "llvm/tools/llvm-exegesis/lib/PerfHelper.h"
+#include "llvm/tools/llvm-exegesis/lib/TargetSelect.h"
+#include "pybind11/detail/common.h"
+#include "pybind11/pybind11.h"
+#include "pybind11_abseil/import_status_module.h"
+#include "pybind11_abseil/status_casters.h"  // IWYU pragma: keep
+
+namespace gematria {
+
+namespace py = ::pybind11;
+
+using namespace llvm;
+using namespace llvm::exegesis;
+
+PYBIND11_MODULE(exegesis_benchmark, m) {
+  m.doc() = "Code for benchmarking basic blocks.";
+
+  py::google::ImportStatusModule();
+
+  py::class_<ExegesisBenchmark>(m, "ExegesisBenchmark")
+      .def("create",
+           []() -> absl::StatusOr<std::unique_ptr<ExegesisBenchmark>> {
+             // LLVM Setup
+             LLVMInitializeX86TargetInfo();
+             LLVMInitializeX86TargetMC();
+             LLVMInitializeX86Target();
+             LLVMInitializeX86AsmPrinter();
+             LLVMInitializeX86AsmParser();
+             LLVMInitializeX86Disassembler();
+
+             // Exegesis Setup
+             InitializeX86ExegesisTarget();
```

Comment:
I'd prefer to put some more protection around this: create a function that calls the init only once, similar to [`InitializeLlvmOnce()`](https://github.com/google/gematria/blob/21db9938a55255a432db6c7f2a2285448de035a6/gematria/llvm/llvm_architecture_support.cc#L37). This guarantees that the initialization is done only once, and there are no race conditions.

I'm not sure these functions can be called multiple times (likely they can), and I'm very sure they are not thread safe (been there, done that, spent a lot of time debugging).